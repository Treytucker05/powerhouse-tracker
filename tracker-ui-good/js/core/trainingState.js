// Placeholder trainingState module to satisfy build. Replace with real implementation when available.
export default {
    init() {
        console.warn('[trainingState placeholder] init() called - no-op');
    },
    getState() { return {}; },
    save() { /* no-op */ }
};
