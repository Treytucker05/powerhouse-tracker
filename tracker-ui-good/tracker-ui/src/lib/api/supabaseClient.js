import { createClient } from '@supabase/supabase-js'

// Environment-based configuration (expected in production build via secrets)
const supabaseUrl = import.meta.env.VITE_SUPABASE_URL || 'https://cqjzvbvmpcqohjarcydg.supabase.co'
const supabaseAnonKey = import.meta.env.VITE_SUPABASE_ANON_KEY || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNxanp2YnZtcGNxb2hqYXJjeWRnIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NDk3NzExNzksImV4cCI6MjA2NTM0NzE3OX0.wioeITJitSKZ9HrZ2iRPmC3xHHj-bL4xDYtT1iXws44'

// Debug logging (safe subset) â€“ remove or gate behind dev flag later if noisy
console.log('[SupabaseClient] mode=env');
console.log('[SupabaseClient] import.meta.env keys:', Object.keys(import.meta.env).filter(k=>k.startsWith('VITE_')));
console.log('[SupabaseClient] URL resolved:', supabaseUrl);
console.log('[SupabaseClient] Key prefix:', supabaseAnonKey.substring(0,12) + '...');

// Lightweight in-memory mock for test environment to avoid any network/filesystem flakiness
export const supabase = (process.env.NODE_ENV === 'test') ? {
    from() {
        return {
            select: () => ({
                eq: () => Promise.resolve({ data: [], error: null })
            })
        };
    },
    auth: {
        getUser: () => ({ data: { user: { id: 'test-user-id' } } }),
        getSession: () => Promise.resolve({ data: { session: null } }),
        onAuthStateChange: (cb) => {
            // Immediately invoke callback once with signed_out to keep logic paths deterministic
            try { cb?.('SIGNED_OUT', { user: null }); } catch { /* noop */ }
            return { data: { subscription: { unsubscribe: () => { } } } };
        },
        signOut: () => Promise.resolve({ error: null })
    }
} : createClient(supabaseUrl, supabaseAnonKey, {
    auth: {
        autoRefreshToken: true,
        persistSession: true
    }
})

// Helper function for user context
export const getCurrentUserId = () => {
    // Return mock user for tests
    if (typeof window === 'undefined' || window?.process?.env?.NODE_ENV === 'test') {
        return 'test-user-id'
    }

    const { data: { user } } = supabase.auth.getUser()
    return user?.id || null
}
