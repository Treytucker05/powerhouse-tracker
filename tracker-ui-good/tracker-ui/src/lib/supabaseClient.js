import { createClient } from '@supabase/supabase-js'

// TEMPORARY HARD-CODE (duplicate helper) - unify later or delete this file in favor of api/supabaseClient.js
const supabaseUrl = 'https://cqjzvbvmpcqohjarcydg.supabase.co'
const supabaseAnonKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNxanp2YnZtcGNxb2hqYXJjeWRnIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NDk3NzExNzksImV4cCI6MjA2NTM0NzE3OX0.wioeITJitSKZ9HrZ2iRPmC3xHHj-bL4xDYtT1iXws44'

console.log('[SupabaseClient:legacy][HARD-CODED TEST] URL:', supabaseUrl)

export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    autoRefreshToken: true,
    persistSession: true
  }
})
