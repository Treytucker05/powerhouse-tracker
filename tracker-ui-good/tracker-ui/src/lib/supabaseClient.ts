import { createClient } from '@supabase/supabase-js';

const url = import.meta.env.VITE_SUPABASE_URL;
const anon = import.meta.env.VITE_SUPABASE_ANON;

if (import.meta.env.DEV) {
    if (!url) throw new Error('[supabase] Missing VITE_SUPABASE_URL');
    if (!anon) throw new Error('[supabase] Missing VITE_SUPABASE_ANON');
}

export const supabase = createClient(url as string, anon as string, {
    auth: {
        autoRefreshToken: true,
        persistSession: true,
        detectSessionInUrl: true
    }
});

export async function getCurrentUserId(): Promise<string | null> {
    const { data: { user } } = await supabase.auth.getUser();
    return user?.id ?? null;
}
