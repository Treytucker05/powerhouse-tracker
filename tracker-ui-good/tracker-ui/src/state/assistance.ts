export type AssistancePerDay = Array<{ day: number; items: Array<{ name: string; category: string }> }>;

export type AssistanceState = {
    assistance?: { perDay: AssistancePerDay };
};

export function setAssistancePerDay(prev: any, perDay: AssistancePerDay) {
    return { ...(prev || {}), assistance: { ...(prev?.assistance || {}), perDay } };
}
