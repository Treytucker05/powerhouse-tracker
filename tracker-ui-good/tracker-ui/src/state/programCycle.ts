// src/state/programCycle.ts
// Single persistence path for finalized 5/3/1 review plan.
// Framework-agnostic; can be swapped out for remote storage later.

export type ActivePlan = {
  id: string;
  createdAt: string;          // ISO timestamp
  cycle: number;              // always 1 on initial create
  reviewState: any;           // opaque ReviewState payload
  nextCycleIncrements: { upper: number; lower: number }; // +5 / +10 (lbs) default
};

const KEY = "ph_active_plan_v1";

export function startCycle(reviewState: any, inc: { upper: number; lower: number } = { upper: 5, lower: 10 }): ActivePlan {
  if (!reviewState?.tm_by_lift || !reviewState?.days) {
    throw new Error("Invalid review state â€” missing TMs or days.");
  }
  const plan: ActivePlan = {
    id: (globalThis as any)?.crypto?.randomUUID?.() ?? String(Date.now()),
    createdAt: new Date().toISOString(),
    cycle: 1,
    reviewState,
    nextCycleIncrements: inc,
  };
  try {
    localStorage.setItem(KEY, JSON.stringify(plan));
  } catch (e) {
    // surface minimal info for debugging
    console.warn("Failed to persist active plan", e);
  }
  return plan;
}

export function getActivePlan(): ActivePlan | null {
  try {
    const raw = localStorage.getItem(KEY);
    if (!raw) return null;
    return JSON.parse(raw) as ActivePlan;
  } catch {
    return null;
  }
}

export function clearActivePlan(): void {
  try { localStorage.removeItem(KEY); } catch { /* noop */ }
}
