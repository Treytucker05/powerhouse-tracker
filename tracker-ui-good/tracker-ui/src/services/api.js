import { supabase, getCurrentUserId } from '../lib/api/supabaseClient';
import { toast } from 'react-toastify';

/**
 * ProgramService - Data abstraction layer for program management
 * Handles saving/loading programs with Supabase + localStorage sync
 */
class ProgramService {
    constructor() {
        this.STORAGE_PREFIX = 'program_';
        this.PROGRAMS_INDEX_KEY = 'programs_index';
    }

    /**
     * Save program to both localStorage and Supabase
     * @param {Object} program - Program data to save
     * @returns {Promise<Object>} - Saved program with updated metadata
     */
    async saveProgram(program) {
        try {
            const userId = getCurrentUserId();
            if (!userId) {
                console.warn('No user ID available for Supabase sync');
            }

            // Validate program data
            if (!program || !program.name) {
                throw new Error('Program name is required');
            }

            // Prepare program data with metadata
            const programData = {
                ...program,
                id: program.id || `program_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                userId: userId || 'anonymous',
                updatedAt: new Date().toISOString(),
                createdAt: program.createdAt || new Date().toISOString(),
                version: (program.version || 0) + 1
            };

            // Save to localStorage first (guaranteed to work)
            const localSaveSuccess = this._saveToLocalStorage(programData);
            if (!localSaveSuccess) {
                throw new Error('Failed to save to local storage');
            }

            // Attempt Supabase save with conflict handling
            let supabaseResult = null;
            if (userId) {
                try {
                    supabaseResult = await this._saveToSupabase(programData, userId);
                } catch (supabaseError) {
                    console.warn('Supabase save failed, continuing with localStorage:', supabaseError);

                    // Handle common conflicts
                    if (supabaseError.code === '23505') { // Unique constraint violation
                        toast.warning('Program name already exists. Using local version.');
                    } else if (supabaseError.code === 'PGRST301') { // Row level security
                        toast.warning('Insufficient permissions. Saved locally only.');
                    } else {
                        toast.warning('Cloud sync failed. Program saved locally.');
                    }
                }
            }

            // Success notification
            const syncStatus = supabaseResult ? 'synced to cloud' : 'saved locally';
            toast.success(`Program "${programData.name}" ${syncStatus} successfully`);

            return {
                ...programData,
                syncedToSupabase: !!supabaseResult,
                lastSyncAt: supabaseResult ? new Date().toISOString() : null
            };

        } catch (error) {
            console.error('Failed to save program:', error);
            toast.error(`Failed to save program: ${error.message}`);
            throw error;
        }
    }

    /**
     * Load program by ID - try Supabase first, fallback to localStorage
     * @param {string} programId - Program ID to load
     * @returns {Promise<Object|null>} - Program data or null if not found
     */
    async loadProgram(programId) {
        try {
            if (!programId) {
                throw new Error('Program ID is required');
            }

            const userId = getCurrentUserId();
            let programData = null;

            // Try Supabase first if user is authenticated
            if (userId) {
                try {
                    programData = await this._loadFromSupabase(programId, userId);
                    if (programData) {
                        // Update localStorage with latest from Supabase
                        this._saveToLocalStorage(programData);
                        return programData;
                    }
                } catch (supabaseError) {
                    console.warn('Supabase load failed, trying localStorage:', supabaseError);
                }
            }

            // Fallback to localStorage
            programData = this._loadFromLocalStorage(programId);
            if (programData) {
                // Check if this is an outdated local copy
                if (userId && !programData.syncedToSupabase) {
                    toast.info('Loading offline version of program');
                }
                return programData;
            }

            // Not found anywhere
            toast.error(`Program "${programId}" not found`);
            return null;

        } catch (error) {
            console.error('Failed to load program:', error);
            toast.error(`Failed to load program: ${error.message}`);
            return null;
        }
    }

    /**
     * Load all programs for the current user
     * @returns {Promise<Array>} - Array of program summaries
     */
    async loadAllPrograms() {
        try {
            const userId = getCurrentUserId();
            let programs = [];

            // Try Supabase first
            if (userId) {
                try {
                    programs = await this._loadAllFromSupabase(userId);
                    if (programs.length > 0) {
                        // Update local index
                        this._updateLocalIndex(programs);
                        return programs;
                    }
                } catch (supabaseError) {
                    console.warn('Supabase load all failed, trying localStorage:', supabaseError);
                }
            }

            // Fallback to localStorage
            programs = this._loadAllFromLocalStorage();
            if (programs.length === 0) {
                toast.info('No programs found');
            }

            return programs;

        } catch (error) {
            console.error('Failed to load programs:', error);
            toast.error('Failed to load programs');
            return [];
        }
    }

    /**
     * Delete program from both storage locations
     * @param {string} programId - Program ID to delete
     * @returns {Promise<boolean>} - Success status
     */
    async deleteProgram(programId) {
        try {
            if (!programId) {
                throw new Error('Program ID is required');
            }

            const userId = getCurrentUserId();
            let success = false;

            // Delete from Supabase first
            if (userId) {
                try {
                    await this._deleteFromSupabase(programId, userId);
                    success = true;
                } catch (supabaseError) {
                    console.warn('Supabase delete failed:', supabaseError);
                }
            }

            // Delete from localStorage
            const localDeleteSuccess = this._deleteFromLocalStorage(programId);

            if (success || localDeleteSuccess) {
                toast.success('Program deleted successfully');
                return true;
            } else {
                throw new Error('Failed to delete from both storage locations');
            }

        } catch (error) {
            console.error('Failed to delete program:', error);
            toast.error(`Failed to delete program: ${error.message}`);
            return false;
        }
    }

    /**
     * Check sync status between local and cloud storage
     * @param {string} programId - Program ID to check
     * @returns {Promise<Object>} - Sync status information
     */
    async checkSyncStatus(programId) {
        try {
            const userId = getCurrentUserId();
            if (!userId) {
                return { synced: false, reason: 'No user authenticated' };
            }

            const localProgram = this._loadFromLocalStorage(programId);
            const cloudProgram = await this._loadFromSupabase(programId, userId);

            if (!localProgram && !cloudProgram) {
                return { synced: false, reason: 'Program not found' };
            }

            if (!localProgram) {
                return { synced: false, reason: 'Missing local copy' };
            }

            if (!cloudProgram) {
                return { synced: false, reason: 'Missing cloud copy' };
            }

            const localVersion = localProgram.version || 0;
            const cloudVersion = cloudProgram.version || 0;

            if (localVersion === cloudVersion) {
                return { synced: true, version: localVersion };
            } else {
                return {
                    synced: false,
                    reason: 'Version mismatch',
                    localVersion,
                    cloudVersion,
                    needsSync: true
                };
            }

        } catch (error) {
            console.error('Failed to check sync status:', error);
            return { synced: false, reason: error.message };
        }
    }

    // Private methods for storage operations

    _saveToLocalStorage(program) {
        try {
            const key = `${this.STORAGE_PREFIX}${program.id}`;
            localStorage.setItem(key, JSON.stringify({
                ...program,
                lastUpdated: new Date().toISOString()
            }));

            // Update programs index
            this._updateLocalIndex([program]);
            return true;
        } catch (error) {
            console.error('LocalStorage save failed:', error);
            return false;
        }
    }

    _loadFromLocalStorage(programId) {
        try {
            const key = `${this.STORAGE_PREFIX}${programId}`;
            const data = localStorage.getItem(key);
            return data ? JSON.parse(data) : null;
        } catch (error) {
            console.error('LocalStorage load failed:', error);
            return null;
        }
    }

    _deleteFromLocalStorage(programId) {
        try {
            const key = `${this.STORAGE_PREFIX}${programId}`;
            localStorage.removeItem(key);

            // Update index
            const index = this._getLocalIndex();
            const updatedIndex = index.filter(p => p.id !== programId);
            localStorage.setItem(this.PROGRAMS_INDEX_KEY, JSON.stringify(updatedIndex));

            return true;
        } catch (error) {
            console.error('LocalStorage delete failed:', error);
            return false;
        }
    }

    _loadAllFromLocalStorage() {
        try {
            const index = this._getLocalIndex();
            return index.map(indexEntry => {
                const program = this._loadFromLocalStorage(indexEntry.id);
                return program || indexEntry; // Fallback to index entry if full program missing
            }).filter(Boolean);
        } catch (error) {
            console.error('LocalStorage load all failed:', error);
            return [];
        }
    }

    async _saveToSupabase(program, userId) {
        const { data, error } = await supabase
            .from('programs')
            .upsert({
                id: program.id,
                user_id: userId,
                name: program.name,
                description: program.description,
                program_type: program.type,
                goal_type: program.goal,
                duration_weeks: program.duration,
                training_days_per_week: program.trainingDaysPerWeek,
                // New enhanced columns for comprehensive program storage
                model: program.trainingModel || program.selectedTrainingModel,
                block_sequence: JSON.stringify(program.blocks || []),
                phases: JSON.stringify(program.phases || []),
                weekly_outline: JSON.stringify(program.weeklyOutline || []),
                parameters: JSON.stringify(program.parameters || {}),
                // Legacy compatibility
                blocks: JSON.stringify(program.blocks || []),
                assessment_id: program.assessmentId,
                total_duration: program.totalDuration,
                is_active: program.isActive,
                version: program.version,
                created_at: program.createdAt,
                updated_at: program.updatedAt
            })
            .select()
            .single();

        if (error) {
            throw error;
        }

        return this._transformSupabaseProgram(data);
    }

    async _loadFromSupabase(programId, userId) {
        const { data, error } = await supabase
            .from('programs')
            .select('*')
            .eq('id', programId)
            .eq('user_id', userId)
            .single();

        if (error) {
            if (error.code === 'PGRST116') { // No rows found
                return null;
            }
            throw error;
        }

        return this._transformSupabaseProgram(data);
    }

    async _loadAllFromSupabase(userId) {
        const { data, error } = await supabase
            .from('programs')
            .select('*')
            .eq('user_id', userId)
            .order('updated_at', { ascending: false });

        if (error) {
            throw error;
        }

        return data.map(this._transformSupabaseProgram);
    }

    async _deleteFromSupabase(programId, userId) {
        const { error } = await supabase
            .from('programs')
            .delete()
            .eq('id', programId)
            .eq('user_id', userId);

        if (error) {
            throw error;
        }
    }

    _transformSupabaseProgram(supabaseData) {
        return {
            id: supabaseData.id,
            userId: supabaseData.user_id,
            name: supabaseData.name,
            description: supabaseData.description,
            type: supabaseData.program_type || supabaseData.type,
            goal: supabaseData.goal_type || supabaseData.goal,
            duration: supabaseData.duration_weeks || supabaseData.duration,
            trainingDaysPerWeek: supabaseData.training_days_per_week,
            // New enhanced fields
            trainingModel: supabaseData.model,
            selectedTrainingModel: supabaseData.model,
            phases: typeof supabaseData.phases === 'string'
                ? JSON.parse(supabaseData.phases)
                : supabaseData.phases,
            weeklyOutline: typeof supabaseData.weekly_outline === 'string'
                ? JSON.parse(supabaseData.weekly_outline)
                : supabaseData.weekly_outline,
            parameters: typeof supabaseData.parameters === 'string'
                ? JSON.parse(supabaseData.parameters)
                : supabaseData.parameters,
            blockSequence: typeof supabaseData.block_sequence === 'string'
                ? JSON.parse(supabaseData.block_sequence)
                : supabaseData.block_sequence,
            // Legacy compatibility
            blocks: typeof supabaseData.blocks === 'string'
                ? JSON.parse(supabaseData.blocks)
                : supabaseData.blocks ||
                (typeof supabaseData.block_sequence === 'string'
                    ? JSON.parse(supabaseData.block_sequence)
                    : supabaseData.block_sequence),
            assessmentId: supabaseData.assessment_id,
            totalDuration: supabaseData.total_duration,
            isActive: supabaseData.is_active,
            version: supabaseData.version,
            createdAt: supabaseData.created_at,
            updatedAt: supabaseData.updated_at,
            syncedToSupabase: true,
            lastSyncAt: new Date().toISOString()
        };
    }

    _getLocalIndex() {
        try {
            const index = localStorage.getItem(this.PROGRAMS_INDEX_KEY);
            return index ? JSON.parse(index) : [];
        } catch (error) {
            console.error('Failed to load local index:', error);
            return [];
        }
    }

    _updateLocalIndex(programs) {
        try {
            const existingIndex = this._getLocalIndex();
            const updatedIndex = [...existingIndex];

            programs.forEach(program => {
                const existingIndex = updatedIndex.findIndex(p => p.id === program.id);
                const indexEntry = {
                    id: program.id,
                    name: program.name,
                    updatedAt: program.updatedAt,
                    createdAt: program.createdAt,
                    type: program.type,
                    goal: program.goal
                };

                if (existingIndex >= 0) {
                    updatedIndex[existingIndex] = indexEntry;
                } else {
                    updatedIndex.push(indexEntry);
                }
            });

            localStorage.setItem(this.PROGRAMS_INDEX_KEY, JSON.stringify(updatedIndex));
        } catch (error) {
            console.error('Failed to update local index:', error);
        }
    }
}

/**
 * Utility functions for easier usage
 */

// Quick save function for React components
export const quickSaveProgram = async (programData, options = {}) => {
    try {
        return await programService.saveProgram(programData);
    } catch (error) {
        if (!options.suppressToast) {
            toast.error(`Save failed: ${error.message}`);
        }
        throw error;
    }
};

// Quick load function for React components  
export const quickLoadProgram = async (programId, options = {}) => {
    try {
        return await programService.loadProgram(programId);
    } catch (error) {
        if (!options.suppressToast) {
            toast.error(`Load failed: ${error.message}`);
        }
        throw error;
    }
};

// Quick load all function for React components
export const quickLoadAllPrograms = async (options = {}) => {
    try {
        return await programService.loadAllPrograms();
    } catch (error) {
        if (!options.suppressToast) {
            toast.error(`Load failed: ${error.message}`);
        }
        throw error;
    }
};

// Quick delete function for React components
export const quickDeleteProgram = async (programId, options = {}) => {
    try {
        return await programService.deleteProgram(programId);
    } catch (error) {
        if (!options.suppressToast) {
            toast.error(`Delete failed: ${error.message}`);
        }
        throw error;
    }
};

// Check if a program is synced with cloud storage
export const checkProgramSyncStatus = async (programId) => {
    return await programService.checkSyncStatus(programId);
};

// Export singleton instance
export const programService = new ProgramService();

// Export the singleton instance as default
export default programService;

// Named exports for convenience
export { ProgramService };
