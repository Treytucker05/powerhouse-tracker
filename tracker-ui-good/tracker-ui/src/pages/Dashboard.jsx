import React from 'react';
import TrainingStatusCard from '../components/dashboard/TrainingStatusCard';
import VolumeTrackingChart from '../components/dashboard/VolumeTrackingChart';
import QuickActions from '../components/dashboard/QuickActions';
import TrainingStatus from '../components/dashboard/TrainingStatus';
// Fallback stubs for components not yet implemented in new structure
const UpcomingSessionsPreview = () => null;
const FatigueRecoveryIndicator = () => null;
import { useDashboardState } from '../lib/state/trainingState';

export default function Dashboard() {
  const {
    currentWeek,
    phase,
    systemicFatigue,
    muscleVolumes,
    refreshDashboard,
  } = useDashboardState();

  return (
    <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
      <div className="space-y-6">
        <TrainingStatusCard currentWeek={currentWeek} phase={phase} />
        <QuickActions onRefresh={refreshDashboard} />
        <UpcomingSessionsPreview />
      </div>
      <div className="space-y-6">
        <VolumeTrackingChart muscleVolumes={muscleVolumes} />
        <TrainingStatus />
        <FatigueRecoveryIndicator systemicFatigue={systemicFatigue} muscleVolumes={muscleVolumes} />
      </div>
    </div>
  );
}
