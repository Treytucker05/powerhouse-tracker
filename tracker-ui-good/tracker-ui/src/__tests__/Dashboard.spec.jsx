import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { supabase } from '../lib/api/supabaseClient'
import Dashboard from '../pages/Dashboard';
import * as state from '../lib/state/trainingState';

// Mocks for heavy child components to keep test deterministic & light
vi.mock('../components/dashboard/VolumeTrackingChart', () => ({ default: () => <div data-testid="volume-chart" /> }))
vi.mock('../components/dashboard/TrainingStatusCard', () => ({ default: () => <div data-testid="status-card" /> }))
vi.mock('../components/dashboard/TrainingStatus', () => ({ default: () => <div data-testid="training-status" /> }))

// Mock actions hook
const startTodaySpy = vi.fn()
vi.mock('../hooks/useQuickActions', () => ({
  useQuickActions: () => ({
    data: {
      startToday: startTodaySpy,
      startTodayLabel: "Start Today's Workout",
      startTodayDisabled: false,
      hasPlannedSession: false
    },
    isLoading: false,
    error: null
  })
}))

// Provide a fresh QueryClient per test file
const qc = new QueryClient({ defaultOptions: { queries: { retry: false } } })

// Helper wrapper
const renderWithProviders = (ui) => render(
  <QueryClientProvider client={qc}>
    <MemoryRouter>
      {ui}
    </MemoryRouter>
  </QueryClientProvider>
)

// Mock the supabase.from() chain to satisfy useWeeklyVolume internal query
const mockSelectChain = {
  eq: () => mockSelectChain,
  gte: () => mockSelectChain,
  lte: () => mockSelectChain,
  order: () => Promise.resolve({ data: [], error: null })
}

vi.spyOn(supabase, 'from').mockReturnValue({ select: () => mockSelectChain })

// Optionally stub visually complex child components if unstable (could mock but leaving for now)

describe('<Dashboard />', () => {
  beforeEach(() => {
    startTodaySpy.mockClear()
    vi.spyOn(state, 'useDashboardState').mockReturnValue({
      currentWeek: 3,
      phase: 'Accumulation',
      systemicFatigue: 0.65,
      muscleVolumes: {
        chest: { current: 14, mev: 10, mrv: 22 },
        back: { current: 18, mev: 12, mrv: 25 },
      },
      refreshDashboard: vi.fn(),
    })
  })

  it('renders core dashboard sections', () => {
    renderWithProviders(<Dashboard />)
    expect(screen.getByText(/Quick Actions/i)).toBeTruthy()
    expect(screen.getByTestId('volume-chart')).toBeTruthy()
    expect(screen.getByTestId('status-card')).toBeTruthy()
  })

  it("invokes start today's workout action", () => {
    renderWithProviders(<Dashboard />)
    fireEvent.click(screen.getByText(/Start Today's Workout/i))
    expect(startTodaySpy).toHaveBeenCalled()
  })
});
