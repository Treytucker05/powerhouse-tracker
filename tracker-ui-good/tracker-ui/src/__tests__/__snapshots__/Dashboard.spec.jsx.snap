// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Dashboard /> > renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 md:grid-cols-2 gap-6"
  >
    <div
      class="space-y-6"
    >
      <div
        class="rounded-lg bg-white dark:bg-gray-800 shadow p-4 flex flex-col"
      >
        <div
          class="text-sm text-gray-500 dark:text-gray-400"
        >
          Current Week
        </div>
        <div
          class="text-2xl font-bold text-gray-900 dark:text-gray-100"
        >
          Week 3
        </div>
        <div
          class="mt-2 text-sm text-gray-500 dark:text-gray-400"
        >
          Phase
        </div>
        <div
          class="text-lg font-semibold text-blue-600 dark:text-blue-400"
        >
          Accumulation
        </div>
        <div
          class="mt-4 text-xs text-gray-400"
        >
          Days to Deload: 
          <span
            class="font-bold text-green-600 dark:text-green-400"
          >
            5
          </span>
        </div>
      </div>
      <div
        class="rounded-lg bg-white dark:bg-gray-800 shadow p-4 flex flex-col gap-3"
      >
        <button
          class="px-4 py-2 rounded bg-blue-600 text-white hover:bg-blue-700"
        >
          Start Todayâ€™s Workout
        </button>
        <button
          class="px-4 py-2 rounded bg-green-600 text-white hover:bg-green-700"
        >
          Log Quick Feedback
        </button>
        <button
          class="px-4 py-2 rounded bg-yellow-500 text-white hover:bg-yellow-600"
        >
          Adjust This Week
        </button>
        <button
          class="px-4 py-2 rounded bg-gray-200 text-gray-800 hover:bg-gray-300 dark:bg-gray-700 dark:text-gray-100 dark:hover:bg-gray-600"
        >
          Refresh
        </button>
      </div>
      <div
        class="rounded-lg bg-white dark:bg-gray-800 shadow p-4"
      >
        <div
          class="font-semibold mb-2 text-gray-900 dark:text-gray-100"
        >
          Upcoming Sessions
        </div>
        <ul
          class="text-sm text-gray-700 dark:text-gray-200"
        >
          <li
            class="mb-1 flex justify-between"
          >
            <span>
              Upper Body Power
            </span>
            <span
              class="text-xs text-gray-400"
            >
              2025-06-27
            </span>
          </li>
          <li
            class="mb-1 flex justify-between"
          >
            <span>
              Lower Body Hypertrophy
            </span>
            <span
              class="text-xs text-gray-400"
            >
              2025-06-29
            </span>
          </li>
          <li
            class="mb-1 flex justify-between"
          >
            <span>
              Full Body Recovery
            </span>
            <span
              class="text-xs text-gray-400"
            >
              2025-07-01
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="space-y-6"
    >
      <div
        class="rounded-lg bg-white dark:bg-gray-800 shadow p-4"
      >
        <div
          class="font-semibold mb-2 text-gray-900 dark:text-gray-100"
        >
          Weekly Volume
        </div>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 220px; min-width: 0;"
        />
      </div>
      <div
        class="rounded-lg bg-white dark:bg-gray-800 shadow p-4 flex flex-col items-center"
      >
        <div
          class="font-semibold mb-2 text-gray-900 dark:text-gray-100"
        >
          Systemic Fatigue
        </div>
        <div
          class="recharts-responsive-container"
          style="width: 120px; height: 120px; min-width: 0;"
        >
          <div
            class="recharts-wrapper"
            style="position: relative; cursor: default; width: 100%; height: 100%; max-height: 120px; max-width: 120px;"
          >
            <svg
              class="recharts-surface"
              cx="50%"
              cy="50%"
              height="120"
              style="width: 100%; height: 100%;"
              viewBox="0 0 120 120"
              width="120"
            >
              <title />
              <desc />
              <defs>
                <clippath
                  id="recharts2-clip"
                >
                  <rect
                    height="110"
                    width="110"
                    x="5"
                    y="5"
                  />
                </clippath>
              </defs>
              <g
                class="recharts-layer recharts-area"
              >
                <g
                  class="recharts-layer recharts-radial-bar-background"
                >
                  <path
                    class="recharts-sector recharts-radial-bar-background-sector"
                    cx="60"
                    cy="60"
                    d="M 60,15.100000000000001
    A 44.9,44.9,0,
    1,1,
    59.9992163471659,15.100000006838663
  L 59.99944323996864,28.10000000485865
            A 31.9,31.9,0,
            1,0,
            60,28.1 Z"
                    fill="#eee"
                    name="Fatigue"
                    role="img"
                  />
                </g>
                <g
                  class="recharts-layer recharts-radial-bar-sectors"
                >
                  <g
                    class="recharts-layer"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
        <div
          class="mt-2 text-lg font-bold text-yellow-500"
        >
          65%
        </div>
        <div
          class="mt-4 w-full"
        >
          <div
            class="text-sm font-semibold text-gray-700 dark:text-gray-200 mb-1"
          >
            Per-Muscle Recovery
          </div>
          <ul
            class="text-xs text-gray-600 dark:text-gray-300"
          >
            <li>
              chest: Working
            </li>
            <li>
              back: Working
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
